libsolidity/analysis/PostTypeChecker.cpp

```diff
@@ -362,6 +365,77 @@ struct NoVariablesInInterfaceChecker: public PostTypeChecker::Checker
 	/// Flag indicating whether we are currently inside a StructDefinition.
 	int m_insideStruct = 0;
 };
+
+struct CircularContractDependencies: public PostTypeChecker::Checker
+{
+	CircularContractDependencies(ErrorReporter& _errorReporter):
+		Checker(_errorReporter)
+	{}
+
+	bool visit(ContractDefinition const& _contract) override
+	{
+		auto joined = boost::join(
+			_contract.annotation().creationCallGraph->get()->createdContracts,
+			_contract.annotation().deployedCallGraph->get()->createdContracts
+		);
+
+		for (auto const& [contract, referencee]: joined)
+			if (contract != &_contract)
+				_contract.annotation().contractDependencies.emplace(contract, referencee);
+
+		return true;
+	}
+
+	void endVisit(ContractDefinition const& _contract) override
+	{
+		auto const& [contract, referencee] = dependenciesAreCyclic(_contract);
+
+		if (contract)
+			m_foundCycles.emplace(contract, referencee);
```

Comment:
What is `contract` here actually? I know it's the thing that `CycleDetector` returns but which node is that? Is that specifically one of the contracts that create/inherit from `_contract` or just random node inside the cycle?

```suggestion
		auto const& [dependentContract, referencee] = dependenciesAreCyclic(_contract);

		if (dependentContract)
			m_foundCycles.emplace(dependentContract, referencee);
```

Reply:
`contract` here is the dependency that causes a cycle, referenced/required by `referencee`. So.. I guess in your words, it is a random node inside the cycle, yes.